app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/LocalFileTransferActivity.java

```diff
@@ -0,0 +1,598 @@
+    return fileReceiverDeviceAddress;
+  }
+
+  public static void copyToOutputStream(@NonNull InputStream inputStream, @NonNull OutputStream outputStream)
+      throws IOException {
+    byte[] bufferForBytes = new byte[1024];
+    int bytesRead;
+
+    Log.d(TAG, "Copying to OutputStream...");
+    while ((bytesRead = inputStream.read(bufferForBytes)) != -1) {
+      outputStream.write(bufferForBytes, 0, bytesRead);
+    }
+
+    outputStream.close();
+    inputStream.close();
+    Log.d(LocalFileTransferActivity.TAG, "Both streams closed");
+  }
+
+  public void changeStatus(int itemIndex, @FileItem.FileStatus int status) {
+    filesToSend.get(itemIndex).setFileStatus(status);
+    fileListAdapter.notifyItemChanged(itemIndex);
+  }
+
+  /* From WifiDirectManager.Callbacks interface */
+  @Override
+  public void updatePeerDevicesList(@NonNull WifiP2pDeviceList peers) {
+    searchingPeersProgressBar.setVisibility(View.GONE);
+    listViewPeerDevices.setVisibility(View.VISIBLE);
+
+    peerDevices.clear();
+    peerDevices.addAll(peers.getDeviceList());
+    ((WifiPeerListAdapter) listViewPeerDevices.getAdapter()).notifyDataSetChanged();
+
+    if (peerDevices.size() == 0) {
+      Log.d(LocalFileTransferActivity.TAG, "No devices found");
+    }
+  }
+
+  @Override
+  public void performHandshakeWithSelectedPeerDevice(@NonNull WifiP2pInfo groupInfo) {
+    if (BuildConfig.DEBUG) {
+      Log.d(TAG, "Starting handshake");
+    }
+    peerGroupHandshakeAsyncTask = new PeerGroupHandshakeAsyncTask(this, groupInfo);
+    peerGroupHandshakeAsyncTask.execute();
+  }
+
+  /* Helper methods used for checking permissions and states of services */
+  private boolean checkCoarseLocationAccessPermission() { // Required by Android to detect wifi-p2p peers
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+
+      if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION)
+          != PackageManager.PERMISSION_GRANTED) {
+
+        if (shouldShowRequestPermissionRationale(Manifest.permission.ACCESS_COARSE_LOCATION)) {
+          alertDialogShower.show(KiwixDialog.LocationPermissionRationale.INSTANCE,
+              new Function0<Unit>() {
+                @Override public Unit invoke() {
+                  requestPermissions(new String[] { Manifest.permission.ACCESS_COARSE_LOCATION },
+                      PERMISSION_REQUEST_CODE_COARSE_LOCATION);
+                  return Unit.INSTANCE;
+                }
+              });
+        } else {
+          requestPermissions(new String[] { Manifest.permission.ACCESS_COARSE_LOCATION },
+              PERMISSION_REQUEST_CODE_COARSE_LOCATION);
+        }
+
+        return false;
+      }
+    }
+
+    return true; // Control reaches here: Either permission granted at install time, or at the time of request
+  }
+
+  private boolean checkExternalStorageWritePermission() { // To access and store the zims
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
```

Comment:
You don't need to do a version check when you use ContextCompat. Context Compatability means that the support(now androidx) libraries will call the appropriate method depending on the api level

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/314ff811fd402d09bab7d5f2d14cae03a8f8e24d#diff-e1d59ae6e15dcb502e125fb6e243d15dR385)